### DISCUSSION TITLE
Queue Progress Indication?

### DISCUSSION THREAD

**Discussion by da3dsoul** [quartznet/quartznet#2256] [Q&A]
The end goal is to indicate to a user that Quartz is doing things and they need to let it cook.
The query I would expect is to select the triggers where the scheduled time is less than now and the state is waiting.
Other than spinning up an ADO connection to query directly, which I don't want to do, because SQLite and concurrency BS, I'm not sure how to accomplish this.

Advice?

**Comment by jafin** [comment]
You could just check if the job is currently running ?

Example https://stackoverflow.com/questions/24282441/how-to-check-whether-quartz-cron-job-is-running

**Reply by da3dsoul** [reply]
I'm not trying to see if a job is running. I'm trying to see if the queue is behind, so to speak. For example, you add 40 jobs set to run at the earliest convenience, then they will take at least a few seconds to run. The user should probably know that the task they need is not processed yet. In this case, it's a few seconds. Realistically, it's more like 20,000 jobs, and they are doing work that takes a few seconds each. I want to know, in the application, that we have 18,000 jobs left in the queue.

**Reply by da3dsoul** [reply]
I suppose I should clarify. I don't want to know that the job I asked for is waiting. As an administrator, I want to know that users are waiting on jobs, and how many

**Reply by jafin** [reply]
Maybe write a process to get all the jobs  and check their next fire time? (pusedocode)

```c#
        foreach (var jobKey in jobKeys)
        {
            var jobDetail = await scheduler.GetJobDetail(jobKey);
            
            // Check if the job is currently in the queue (not yet executed)
            var triggers = await scheduler.GetTriggersOfJob(jobKey);
            var nextFireTime = triggers.FirstOrDefault()?.GetNextFireTimeUtc();

            if (nextFireTime.HasValue && nextFireTime > DateTime.UtcNow)
            {
               //add to collection of queued jobs
            }
        }
```

But for 20,000 jobs, unsure what impact this has on performance depending on how often you execute..

**Reply by da3dsoul** [reply]
Yeah... That's probably going to be too inefficient. I guess the answer is that I need to use SQL.
I find it odd that there aren't more administration features, considering that you wouldn't need the kind of performance and scalability of Quartz unless you have so many jobs that you might want to administer them.

**Reply by jafin** [reply]
TBH, The SQL you write will probably end up being like the code mentioned, you're going to have to find next fire time of jobs to determine if they are queued AFAIK.  

I feel like your usage of Quartz is more aligned to a queueing system vs a job scheduling system..  Certainly, it can do it, but if you are after a queueing process (i.e. run at the next available opportunity) there may be alternative choices.

**Reply by da3dsoul** [reply]
The only other reasonable alternative I've found is HangFire, and that's not really a good choice for open source projects, as the licensing is...not free. If you have suggestions, I'd love to hear them
